Logging package

matkrak opened the pull request:
Should be more logging.info?
i.e. in <b>array_finalize</b> ?

And this exception I rised, is that OK ?

Also I am adding loggin info to profile.py now.
Should I also 

```
import logging
logging.basicConfig(level=logging.ERROR)
```

there or I should rather connect this two loggers somehow?

matkrak committed:
First attempt

Should be more logging.info?
i.e. in <b>array_finalize</b> ?

And this exception I rised, is that OK ?

grzanka commented on beprof/curve.py:
```diff
@@ -43,6 +45,7 @@ class Curve(np.ndarray):
     """
 
     def __new__(cls, input_array, **meta):
+        logging.info('Creating Curve object')
```
Be more informative. Say sth about parameters of the method. It would be useful to know if `meta` is empty or what does it store

grzanka commented on beprof/curve.py:
```diff
@@ -152,6 +159,7 @@ def __str__(self):
 
 
 def main():
+    logging.info('Inside main() function')
```
This is useless

grzanka commented on beprof/curve.py:
```diff
@@ -123,6 +129,7 @@ def rebinned(self, step=0.1, fixp=0):
         :param fixp: fixed point one of the points in new domain
         :return: new Curve object with domain specified by step and fixp parameters
         '''
+        logging.info('Inside rebinned() method')
```
Add some info about parameters - it will help you do later debugging

matkrak committed:
Better info provided 

Also changing logging level just for testing purpose.

matkrak commented:
And sorry for working with not-up-to-date repo, but I started before you merged #16 and #18. Hope to make it clear tomorrow on Skype how should I work to not have this situations. For now I am just gonna ignore things like hardcoded meta in classes

matkrak committed:
Added logging info

Need to add more but my question is in PR

matkrak commented:
1) About rising exceptions. You didn't say anything about the way I did it, but I see you did something different in Profile.normalize(). Should I stick to it? Also I must add logging.error or logging.exception there, and if it should be logging.exception I should change one in curve.change_domain() as well.
2) Should I add logging info to every method, even if they just can't fail ? Like Profile.fwhm() or Profile.width()?
3) You mentioned that we assume the shape is correct. Can you name 1 place where I should make changes? Then I will make them in all files.

grzanka commented on beprof/curve.py:
```diff
@@ -43,6 +45,7 @@ class Curve(np.ndarray):
     """
 
     def __new__(cls, input_array, **meta):
+        logging.info('Creating Curve object, metadata is: %s', meta)
```
You are using Python2 string formatting. In Python3 it has been replaced with new one. Why was it done is described here: https://www.python.org/dev/peps/pep-3101/ . How to use it, read this: https://docs.python.org/3.5/library/string.html#formatstrings

grzanka commented on beprof/curve.py:
```diff
@@ -95,15 +98,19 @@ def change_domain(self, domain):
+        try:
+            if np.max(domain) > np.max(self.x) or np.min(domain) < np.min(self.x):
+                raise IndexError
+        except IndexError as ie:
```
It doesn't make sense to raise an exception and catch it immediately. Try to google: "Python exception best practices" and read a bit about them.

grzanka commented:
Ad1) Replied in the code.
Ad2) Add logging on debug/info level wherever you think it might help debugging the code.
Ad3) Creating curve object.

matkrak committed:
New  version

1)Changed formatting text in logging
2)Changed exception thrown in change_domain

shape variable in Curve.__new__ will be used to check whether input_array has good dimensions.

matkrak committed:
Changed formatting

matkrak commented:
About this shape validation. Should I put in `Curve.__new__` code like

```
if shape[1] != 2 # could be 3d in the future then also : and shape[1] != 3
    logging.error('Creating new Curve object failed - invalid shape of input_array)
    raise Exception(' Invalid shape of input_array')
```

or how should I handle it?
Also when I am throwing an exception I probably should change logging.error() to logging.exception(), is that right?

grzanka commented:
Reporting that shape is invalid doesn't tell user how to fix the problem. User for sure doesn't know whats the correct shape. It would be also nice to tell him what is the actual shape of his object.

Second question. Have you read carefully documentation about logging.error() and logging.exception() ? I mean:
- https://docs.python.org/3.5/howto/logging.html
- https://docs.python.org/3.5/library/logging.html
  Typing in your browser "when to use logging exception" might also help you.

You should also improve your skills in asking questions. Try to figure out what RTFM means.
Read "Before you ask" here: http://www.catb.org/esr/faqs/smart-questions.html (and https://github.com/galvao/RTFM or others that appear in google when you type "how to ask smart questions").

matkrak committed:
Checking shape of input_array

I am a bit confused about this exceptions thing. 
I decided to raise the exception and not to provide except block so that program will stop when condition in if statement is true. Is that good approach?

matkrak commented:
The right question here might be:
what do you want to be done, if the shape of input_array is not correct? 
2 of many possible options are:
1) log error message and return None
2) log error message and rise an exception

grzanka commented:
Try to study Quantity class implemented here:
https://github.com/NatalieP-J/astropy/blob/master/astropy/units/quantity.py

matkrak committed:
Updated docstring exceptions

grzanka commented on beprof/curve.py:
```diff
@@ -95,15 +111,21 @@ def change_domain(self, domain):
+                          'New domain range: [{2}, {3}]'.format(np.min(self.x), np.max(self.x),
+                                                                np.min(domain), np.max(domain))
+                          )
+            raise ValueError('The old domain does not include the new one!\n'
```
Text duplication. Violates DRY

matkrak committed:
Added some logging messages

Shape control provided in Curve class as menioned in the comments.
Due to super() Curve.__new__ is going to be called every time Pofile or LateralProfile
objects are created.

matkrak committed:
Modified some logs 

Not repeating myself this time

matkrak committed:
Changed logging level

matkrak committed:
still cleaning the mess

matkrak commented:
Ready to merge

grzanka commented on beprof/curve.py:
```diff
@@ -169,8 +189,17 @@ def main():
+=======
     k2 = k.view(np.ndarray)
     print(k2)
+>>>>>>> bcdae8cdcaa6a29186731842b238ac252b29900e
```
What is that ? Unresolved conflict ?

grzanka commented on beprof/profile.py:
```diff
@@ -10,8 +11,19 @@ class Profile(Curve):
+=======
     def __new__(cls, input_array, **kwargs):
         new = super().__new__(cls, input_array, **kwargs)
+>>>>>>> bcdae8cdcaa6a29186731842b238ac252b29900e
```
Unresolved conflict again ?

matkrak committed:
removed conflict

matkrak committed:
conflict removed

I see that I probably pushed wrong files due to all this mess with duplicated beprof directory
